{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "da485b0a",
   "metadata": {},
   "source": [
    "# Data analysis with TogoDX: Network visualization\n",
    "\n",
    "TogoDX/Human (https://togodx.dbcls.jp/human) の生物学データ解析への活用例。文責: 大田 (DBCLS)}\n",
    "\n",
    "## Background\n",
    "\n",
    "TogoDX は、 RDF 化され知識グラフとして統合された生命医科学データにアクセスするためのインターフェースである。ユーザはデータセット全体、あるいはユーザがIDのリストとして与えたサブセットに対して、統合された属性を用いてフィルタリング (filter) や属性情報の追加 (map attributes) を行い、次のアクションの対象となるデータセットを抽出することができる。\n",
    "\n",
    "ただし、TogoDX によって出力されたデータは primary dataset (ユーザが選択する統合の軸となるデータセット) をハブとして複数の属性が結合された星型グラフの集合である。出力テーブルにおける行間（複数の星型グラフ）やカラム間 (星型グラフにおける葉ノード間）の結合は保証されていない。抽出したデータはロング型の TSV として出力・ダウンロードが可能である。\n",
    "\n",
    "## User story\n",
    "\n",
    "- ユーザはベンチ・サイエンティストである。何かしらの分子を定量する実験を行い、群間比較を行う。結果として興味のある分子のリストを得る。\n",
    "- リストに含まれる分子はそれぞれ何かしらの定量値を持っているが、分子間の関係はわからない。\n",
    "- TogoDX を用いて得られた分子のリストに知識グラフの情報を付加する。\n",
    "- それぞれが独立したリスト形式のデータに対してネットワーク構造を持つ知識グラフの情報を与えることで、グラフとして扱うことができる。\n",
    "- グラフの状態で可視化やグラフ分析を行うことで、ハブとなる分子、クラスターを構成する分子群などを捉えることで、分子間の関係性を説明する。\n",
    "\n",
    "## Procedure\n",
    "\n",
    "### TogoDX での操作\n",
    "\n",
    "- TogoDX/Human に分子IDのリストを入力し map your IDs を実行する。\n",
    "    - 入力されたIDに対応するエントリと、そこに接続するデータセットの属性にピンが刺さる。\n",
    "- (optional) リストのIDの数が多い場合には Add filters 機能を実行する。\n",
    "    - たとえば gene list に対して chromosome や gene type で絞り込む。\n",
    "- Map attributes 機能で、与えたIDに対して興味のある属性情報を追加する。\n",
    "    - たとえば gene list に対して protein structure, pathway, drug などの属性を追加する。\n",
    "- View results で得られたデータを確認する。問題がなければ TSV をダウンロードする。\n",
    "    - あまりに数が多いと後の解析が大変なので戻ってフィルタで減らすか覚悟を決める。\n",
    "\n",
    "### Notebook での操作\n",
    "\n",
    "- 分析に必要なパッケージ群をインストールし、ロードする。\n",
    "- TogoDX からダウンロードしたTSVを読み込む。\n",
    "    - 統計値などを取得し読み込んだテーブルの規模や属性の分布を把握する。\n",
    "- (optional) 前処理を行う。\n",
    "    - ダウンロードした TSV の1行はハブとなる primary dataset の1エントリと、それに対応する属性を持つ1エントリの 1-1 ペアである。\n",
    "    - 下の例では、その後操作しやすいように行を識別する pair ID を生成するなどの簡単な前処理を行っている。\n",
    "- TogoID を用いて星型グラフにおける葉ノード間を接続する。\n",
    "    - TogoID API を利用して、葉ノード同士を接続するリンクを取得、グラフに追加する。\n",
    "- 得られたグラフを可視化する。\n",
    "\n",
    "## Further analysis\n",
    "\n",
    "可視化されたグラフに対しては解析者のアイディア次第で様々な解析が可能である。例として以下のようなアクションが想定される。\n",
    "\n",
    "- サブグラフの規模によってフィルタし、独立したグラフに存在する分子を特定する。\n",
    "- 追加された属性に対して任意の weight を与えることで、より重要なパスを特定し、そこに関連する分子を特定する。\n",
    "- グラフの中心性を求める手法を用いてハブとなる分子を特定する。\n",
    "\n",
    "### Limitation\n",
    "\n",
    "上で想定する解析における既知の問題は以下。\n",
    "\n",
    "- ダウンロードされた TSV において target dataset のラベル情報が欠けている。\n",
    "    - TSV生成の実装における問題なので技術的に解消が可能。\n",
    "- TogoID API のスケーラビリティ上限が未知\n",
    "    - 非常に短い間隔で叩き続けるとレスポンスが返ってこない可能性がある。\n",
    "    - 下に示した実装の例は非効率なので、改善できる可能性がある。\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd7fbe8a",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "961bfecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from functools import reduce\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1e371f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_rows', 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6d767051",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from urllib.request import urlopen\n",
    "import json\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "417e51aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "!{sys.executable} -m pip install pyyaml --quiet\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3ba3b2e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "!{sys.executable} -m pip install pyvis --quiet\n",
    "# https://pyvis.readthedocs.io/\n",
    "from pyvis.network import Network"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c685f4a",
   "metadata": {},
   "source": [
    "## Load dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "15aabfbc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_dataset</th>\n",
       "      <th>orig_entry</th>\n",
       "      <th>orig_label</th>\n",
       "      <th>dest_dataset</th>\n",
       "      <th>dest_entry</th>\n",
       "      <th>node</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000379939</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000400445</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000537702</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000629018</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>26960</td>\n",
       "      <td>gene_evolutionary_conservation_homologene</td>\n",
       "      <td>Insect, Worm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4529</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>P12755</td>\n",
       "      <td>protein_disease_related_proteins_uniprot</td>\n",
       "      <td>Disease variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4530</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>P12755</td>\n",
       "      <td>structure_data_existence_uniprot</td>\n",
       "      <td>Proteins with structure data</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4531</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>P12755</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>Signaling Pathways</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4532</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>P12755</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>Gene expression (Transcription)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4533</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000157933</td>\n",
       "      <td>SKI</td>\n",
       "      <td>mondo</td>\n",
       "      <td>0008426</td>\n",
       "      <td>disease_diseases_mondo</td>\n",
       "      <td>human disease or disorder</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4534 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      orig_dataset       orig_entry orig_label        dest_dataset  \\\n",
       "0     ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "1     ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "2     ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "3     ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "4     ensembl_gene  ENSG00000172915       NBEA            ncbigene   \n",
       "...            ...              ...        ...                 ...   \n",
       "4529  ensembl_gene  ENSG00000157933        SKI             uniprot   \n",
       "4530  ensembl_gene  ENSG00000157933        SKI             uniprot   \n",
       "4531  ensembl_gene  ENSG00000157933        SKI             uniprot   \n",
       "4532  ensembl_gene  ENSG00000157933        SKI             uniprot   \n",
       "4533  ensembl_gene  ENSG00000157933        SKI               mondo   \n",
       "\n",
       "           dest_entry                                       node  \\\n",
       "0     ENST00000379939                 transcript_biotype_ensembl   \n",
       "1     ENST00000400445                 transcript_biotype_ensembl   \n",
       "2     ENST00000537702                 transcript_biotype_ensembl   \n",
       "3     ENST00000629018                 transcript_biotype_ensembl   \n",
       "4               26960  gene_evolutionary_conservation_homologene   \n",
       "...               ...                                        ...   \n",
       "4529           P12755   protein_disease_related_proteins_uniprot   \n",
       "4530           P12755           structure_data_existence_uniprot   \n",
       "4531           P12755   interaction_proteins_in_pathway_reactome   \n",
       "4532           P12755   interaction_proteins_in_pathway_reactome   \n",
       "4533          0008426                     disease_diseases_mondo   \n",
       "\n",
       "                                value  \n",
       "0                      protein coding  \n",
       "1                      protein coding  \n",
       "2                      protein coding  \n",
       "3                      protein coding  \n",
       "4                        Insect, Worm  \n",
       "...                               ...  \n",
       "4529                  Disease variant  \n",
       "4530     Proteins with structure data  \n",
       "4531               Signaling Pathways  \n",
       "4532  Gene expression (Transcription)  \n",
       "4533        human disease or disorder  \n",
       "\n",
       "[4534 rows x 7 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path = \"../data/togodx-20230217-12655.tsv\"\n",
    "d = pd.read_table(data_path)\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ef1b4fbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_dataset</th>\n",
       "      <th>orig_entry</th>\n",
       "      <th>orig_label</th>\n",
       "      <th>dest_dataset</th>\n",
       "      <th>dest_entry</th>\n",
       "      <th>node</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4534</td>\n",
       "      <td>4534</td>\n",
       "      <td>4534</td>\n",
       "      <td>4534</td>\n",
       "      <td>4534</td>\n",
       "      <td>4534</td>\n",
       "      <td>4534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>1</td>\n",
       "      <td>154</td>\n",
       "      <td>154</td>\n",
       "      <td>4</td>\n",
       "      <td>1894</td>\n",
       "      <td>8</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000044115</td>\n",
       "      <td>CTNNA1</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>836</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>protein coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>4534</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>2528</td>\n",
       "      <td>19</td>\n",
       "      <td>903</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        orig_dataset       orig_entry orig_label dest_dataset dest_entry  \\\n",
       "count           4534             4534       4534         4534       4534   \n",
       "unique             1              154        154            4       1894   \n",
       "top     ensembl_gene  ENSG00000044115     CTNNA1      uniprot        836   \n",
       "freq            4534              126        126         2528         19   \n",
       "\n",
       "                                            node           value  \n",
       "count                                       4534            4534  \n",
       "unique                                         8              99  \n",
       "top     interaction_proteins_in_pathway_reactome  protein coding  \n",
       "freq                                         903             885  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c771f412",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List up the selected attributes\n",
    "# d['node'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13fa9f01",
   "metadata": {},
   "source": [
    "## Extract subset\n",
    "\n",
    "手法の確認のためにまず１行目だけを取り出して可視化する。つまり、星型グラフ１個について情報を付加する。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fa85e578",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000172915'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_row_orig_id = d.head(1)['orig_entry'][0]\n",
    "first_row_orig_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7f73f6bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_82/1901048836.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  subset['orig_id'] = subset['orig_dataset'] + ':' + subset['orig_entry']\n",
      "/tmp/ipykernel_82/1901048836.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  subset['dest_id'] = subset['dest_dataset'] + ':' + subset['dest_entry']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_dataset</th>\n",
       "      <th>orig_entry</th>\n",
       "      <th>orig_label</th>\n",
       "      <th>dest_dataset</th>\n",
       "      <th>dest_entry</th>\n",
       "      <th>node</th>\n",
       "      <th>value</th>\n",
       "      <th>orig_id</th>\n",
       "      <th>dest_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000379939</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000379939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000400445</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000400445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000537702</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000537702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ensembl_transcript</td>\n",
       "      <td>ENST00000629018</td>\n",
       "      <td>transcript_biotype_ensembl</td>\n",
       "      <td>protein coding</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000629018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>26960</td>\n",
       "      <td>gene_evolutionary_conservation_homologene</td>\n",
       "      <td>Insect, Worm</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ncbigene:26960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>A0A8I5QKR1</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>Not in the pathway</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5QKR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>A0A8I5QKR6</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>Not in the pathway</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5QKR6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>Q5T321</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>Not in the pathway</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:Q5T321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>uniprot</td>\n",
       "      <td>Q8NFP9</td>\n",
       "      <td>interaction_proteins_in_pathway_reactome</td>\n",
       "      <td>Not in the pathway</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:Q8NFP9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ensembl_gene</td>\n",
       "      <td>ENSG00000172915</td>\n",
       "      <td>NBEA</td>\n",
       "      <td>mondo</td>\n",
       "      <td>0030930</td>\n",
       "      <td>disease_diseases_mondo</td>\n",
       "      <td>human disease or disorder</td>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>mondo:0030930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    orig_dataset       orig_entry orig_label        dest_dataset  \\\n",
       "0   ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "1   ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "2   ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "3   ensembl_gene  ENSG00000172915       NBEA  ensembl_transcript   \n",
       "4   ensembl_gene  ENSG00000172915       NBEA            ncbigene   \n",
       "..           ...              ...        ...                 ...   \n",
       "75  ensembl_gene  ENSG00000172915       NBEA             uniprot   \n",
       "76  ensembl_gene  ENSG00000172915       NBEA             uniprot   \n",
       "77  ensembl_gene  ENSG00000172915       NBEA             uniprot   \n",
       "78  ensembl_gene  ENSG00000172915       NBEA             uniprot   \n",
       "79  ensembl_gene  ENSG00000172915       NBEA               mondo   \n",
       "\n",
       "         dest_entry                                       node  \\\n",
       "0   ENST00000379939                 transcript_biotype_ensembl   \n",
       "1   ENST00000400445                 transcript_biotype_ensembl   \n",
       "2   ENST00000537702                 transcript_biotype_ensembl   \n",
       "3   ENST00000629018                 transcript_biotype_ensembl   \n",
       "4             26960  gene_evolutionary_conservation_homologene   \n",
       "..              ...                                        ...   \n",
       "75       A0A8I5QKR1   interaction_proteins_in_pathway_reactome   \n",
       "76       A0A8I5QKR6   interaction_proteins_in_pathway_reactome   \n",
       "77           Q5T321   interaction_proteins_in_pathway_reactome   \n",
       "78           Q8NFP9   interaction_proteins_in_pathway_reactome   \n",
       "79          0030930                     disease_diseases_mondo   \n",
       "\n",
       "                        value                       orig_id  \\\n",
       "0              protein coding  ensembl_gene:ENSG00000172915   \n",
       "1              protein coding  ensembl_gene:ENSG00000172915   \n",
       "2              protein coding  ensembl_gene:ENSG00000172915   \n",
       "3              protein coding  ensembl_gene:ENSG00000172915   \n",
       "4                Insect, Worm  ensembl_gene:ENSG00000172915   \n",
       "..                        ...                           ...   \n",
       "75         Not in the pathway  ensembl_gene:ENSG00000172915   \n",
       "76         Not in the pathway  ensembl_gene:ENSG00000172915   \n",
       "77         Not in the pathway  ensembl_gene:ENSG00000172915   \n",
       "78         Not in the pathway  ensembl_gene:ENSG00000172915   \n",
       "79  human disease or disorder  ensembl_gene:ENSG00000172915   \n",
       "\n",
       "                               dest_id  \n",
       "0   ensembl_transcript:ENST00000379939  \n",
       "1   ensembl_transcript:ENST00000400445  \n",
       "2   ensembl_transcript:ENST00000537702  \n",
       "3   ensembl_transcript:ENST00000629018  \n",
       "4                       ncbigene:26960  \n",
       "..                                 ...  \n",
       "75                  uniprot:A0A8I5QKR1  \n",
       "76                  uniprot:A0A8I5QKR6  \n",
       "77                      uniprot:Q5T321  \n",
       "78                      uniprot:Q8NFP9  \n",
       "79                       mondo:0030930  \n",
       "\n",
       "[80 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orig_id = first_row_orig_id\n",
    "subset = d[d['orig_entry'] == orig_id]\n",
    "# Add columns of identifiers in dataset:id form\n",
    "subset['orig_id'] = subset['orig_dataset'] + ':' + subset['orig_entry']\n",
    "subset['dest_id'] = subset['dest_dataset'] + ':' + subset['dest_entry']\n",
    "subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "593aaaf7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_id</th>\n",
       "      <th>dest_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000379939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000400445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000537702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ensembl_transcript:ENST00000629018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>ncbigene:26960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A0D9SF28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KQL6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KQP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KRX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KRZ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KS18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KS58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KSX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KTV6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KTY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KWL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KX92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KXL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KXR7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KXW8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KXY8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5KZ13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5QJB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5QJQ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5QKR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:A0A8I5QKR6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:Q5T321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>uniprot:Q8NFP9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ensembl_gene:ENSG00000172915</td>\n",
       "      <td>mondo:0030930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         orig_id                             dest_id\n",
       "0   ensembl_gene:ENSG00000172915  ensembl_transcript:ENST00000379939\n",
       "1   ensembl_gene:ENSG00000172915  ensembl_transcript:ENST00000400445\n",
       "2   ensembl_gene:ENSG00000172915  ensembl_transcript:ENST00000537702\n",
       "3   ensembl_gene:ENSG00000172915  ensembl_transcript:ENST00000629018\n",
       "4   ensembl_gene:ENSG00000172915                      ncbigene:26960\n",
       "8   ensembl_gene:ENSG00000172915                  uniprot:A0A0D9SF28\n",
       "9   ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KQL6\n",
       "10  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KQP5\n",
       "11  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KRX1\n",
       "12  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KRZ1\n",
       "13  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KS18\n",
       "14  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KS58\n",
       "15  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KSX3\n",
       "16  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KTV6\n",
       "17  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KTY1\n",
       "18  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KWL1\n",
       "19  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KX92\n",
       "20  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KXL4\n",
       "21  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KXR7\n",
       "22  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KXW8\n",
       "23  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KXY8\n",
       "24  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5KZ13\n",
       "25  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5QJB2\n",
       "26  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5QJQ0\n",
       "27  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5QKR1\n",
       "28  ensembl_gene:ENSG00000172915                  uniprot:A0A8I5QKR6\n",
       "29  ensembl_gene:ENSG00000172915                      uniprot:Q5T321\n",
       "30  ensembl_gene:ENSG00000172915                      uniprot:Q8NFP9\n",
       "79  ensembl_gene:ENSG00000172915                       mondo:0030930"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate ID pairs that connect each other\n",
    "id_pairs = subset[['orig_id', 'dest_id']].drop_duplicates()\n",
    "id_pairs = id_pairs[id_pairs['orig_id'] != id_pairs['dest_id']]\n",
    "id_pairs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb5c7c1d",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd5e101",
   "metadata": {},
   "source": [
    "### Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "846f7c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "attributes_json_url = 'https://github.com/togodx/togodx-config-human/raw/develop/config/attributes.dx-server.json'\n",
    "attributes_json = json.loads(urlopen(attributes_json_url).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "19f6959e",
   "metadata": {},
   "outputs": [],
   "source": [
    "togoid_dataset_config_url = 'https://github.com/togoid/togoid-config/raw/main/config/dataset.yaml'\n",
    "dataset_config = yaml.safe_load(urlopen(togoid_dataset_config_url).read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c8fe679b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'affy_probeset': '#53c666',\n",
       " 'bioproject': '#696969',\n",
       " 'biosample': '#696969',\n",
       " 'ccds': '#53c666',\n",
       " 'chebi': '#a853c6',\n",
       " 'chembl_compound': '#a853c6',\n",
       " 'chembl_target': '#a2c653',\n",
       " 'civic_gene': '#53c666',\n",
       " 'clinvar': '#53c3c6',\n",
       " 'dbsnp': '#53c3c6',\n",
       " 'dgidb': '#c65381',\n",
       " 'doid': '#5361c6',\n",
       " 'drugbank': '#a853c6',\n",
       " 'ec': '#696969',\n",
       " 'ena': '#53c666',\n",
       " 'ensembl_gene': '#53c666',\n",
       " 'ensembl_protein': '#a2c653',\n",
       " 'ensembl_transcript': '#53c666',\n",
       " 'flybase_gene': '#53c666',\n",
       " 'gea': '#696969',\n",
       " 'glytoucan': '#673aa6',\n",
       " 'go': '#696969',\n",
       " 'hgnc': '#53c666',\n",
       " 'hgnc_symbol': '#53c666',\n",
       " 'hmdb': '#a853c6',\n",
       " 'homologene': '#53c666',\n",
       " 'cog': '#53c666',\n",
       " 'hp': '#5361c6',\n",
       " 'human_protein_atlas': '#a2c653',\n",
       " 'inchi_key': '#a853c6',\n",
       " 'insdc': '#53c666',\n",
       " 'insdc_master': '#696969',\n",
       " 'intact': '#c65381',\n",
       " 'interpro': '#a2c653',\n",
       " 'kegg_compound': '#a853c6',\n",
       " 'kegg_disease': '#5361c6',\n",
       " 'kegg_orthology': '#53c666',\n",
       " 'kegg_pathway': '#c65381',\n",
       " 'kegg_reaction': '#c65381',\n",
       " 'lrg': '#53c666',\n",
       " 'mbgd_gene': '#53c666',\n",
       " 'meddra': '#5361c6',\n",
       " 'medgen': '#5361c6',\n",
       " 'mesh': '#5361c6',\n",
       " 'mgi': '#53c666',\n",
       " 'mirbase': '#53c666',\n",
       " 'mondo': '#5361c6',\n",
       " 'nando': '#5361c6',\n",
       " 'ncbigene': '#53c666',\n",
       " 'ncbiprotein': '#a2c653',\n",
       " 'ncit_disease': '#5361c6',\n",
       " 'oma_group': '#53c666',\n",
       " 'oma_protein': '#a2c653',\n",
       " 'omim_gene': '#53c666',\n",
       " 'omim_phenotype': '#5361c6',\n",
       " 'orphanet': '#5361c6',\n",
       " 'pdb': '#c68753',\n",
       " 'pdb_ccd': '#a853c6',\n",
       " 'pfam': '#a2c653',\n",
       " 'pubchem_compound': '#a853c6',\n",
       " 'pubchem_substance': '#a853c6',\n",
       " 'pubchem_pathway': '#c65381',\n",
       " 'pubmed': '#696969',\n",
       " 'reactome_pathway': '#c65381',\n",
       " 'reactome_reaction': '#c65381',\n",
       " 'refseq_genomic': '#53c666',\n",
       " 'refseq_protein': '#a2c653',\n",
       " 'refseq_rna': '#53c666',\n",
       " 'rgd': '#53c666',\n",
       " 'rhea': '#c65381',\n",
       " 'sgd': '#53c666',\n",
       " 'sra_accession': '#696969',\n",
       " 'sra_analysis': '#696969',\n",
       " 'sra_experiment': '#696969',\n",
       " 'sra_project': '#696969',\n",
       " 'sra_run': '#696969',\n",
       " 'sra_sample': '#696969',\n",
       " 'tair': '#53c666',\n",
       " 'togovar': '#53c3c6',\n",
       " 'uniprot': '#a2c653',\n",
       " 'uniprot_mnemonic': '#a2c653',\n",
       " 'vgnc': '#53c666',\n",
       " 'wikipathways': '#c65381',\n",
       " 'wormbase_gene': '#53c666',\n",
       " 'xenbase_gene': '#53c666',\n",
       " 'zfin_gene': '#53c666'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Color code\n",
    "categories_color = {\n",
    "  \"Analysis\": { \"color\": \"#696969\" },\n",
    "  \"Compound\": { \"color\": \"#a853c6\" },\n",
    "  \"Disease\": { \"color\": \"#5361c6\" },\n",
    "  \"Domain\": { \"color\": \"#a2c653\" },\n",
    "  \"Experiment\": { \"color\": \"#696969\" },\n",
    "  \"Function\": { \"color\": \"#696969\" },\n",
    "  \"Gene\": { \"color\": \"#53c666\" },\n",
    "  \"Glycan\": { \"color\": \"#673aa6\" },\n",
    "  \"Interaction\": { \"color\": \"#c65381\" },\n",
    "  \"Literature\": { \"color\": \"#696969\" },\n",
    "  \"Ortholog\": { \"color\": \"#53c666\" },\n",
    "  \"Pathway\": { \"color\": \"#c65381\" },\n",
    "  \"Probe\": { \"color\": \"#53c666\" },\n",
    "  \"Project\": { \"color\": \"#696969\" },\n",
    "  \"Protein\": { \"color\": \"#a2c653\" },\n",
    "  \"Reaction\": { \"color\": \"#c65381\" },\n",
    "  \"Sample\": { \"color\": \"#696969\" },\n",
    "  \"SequenceRun\": { \"color\": \"#696969\" },\n",
    "  \"Structure\": { \"color\": \"#c68753\" },\n",
    "  \"Submission\": { \"color\": \"#696969\" },\n",
    "  \"Taxonomy\": { \"color\": \"#006400\" },\n",
    "  \"Transcript\": { \"color\": \"#53c666\" },\n",
    "  \"Variant\": { \"color\": \"#53c3c6\" },\n",
    "};\n",
    "\n",
    "dataset_color = {}\n",
    "for ds in dataset_config:\n",
    "    cat = dataset_config[ds]['category']\n",
    "    if not cat in categories_color:\n",
    "        continue\n",
    "    dataset_color[ds] = categories_color[cat]['color']\n",
    "\n",
    "dataset_color"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ccfa758",
   "metadata": {},
   "source": [
    "### PyVis init"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b3191a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "net = Network(notebook=True, cdn_resources='in_line', bgcolor='gray', font_color='white')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "684c5370",
   "metadata": {},
   "source": [
    "### Add hub node (primary dataset) and connected nodes (attributes added by togodx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c482c98e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for oid in id_pairs['orig_id'].drop_duplicates():\n",
    "    net.add_node(oid, color='white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4452f4ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "for did in id_pairs['dest_id'].drop_duplicates():\n",
    "    ds = did.split(':')[0]\n",
    "    color = dataset_color[ds]\n",
    "    net.add_node(did, color=color)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3cadaa83",
   "metadata": {},
   "outputs": [],
   "source": [
    "net.add_edges(id_pairs.to_numpy())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1549313e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"600px\"\n",
       "            src=\"mygraph.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x40792050a0>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net.toggle_physics(True)\n",
    "net.show('mygraph.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c4b4f685",
   "metadata": {},
   "outputs": [],
   "source": [
    "outer_entries = subset[subset['dest_dataset'] != subset['orig_dataset'][0] ][['dest_dataset', 'dest_entry']].drop_duplicates()\n",
    "# outer_entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b18cb338",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets_dict = attributes_json['datasets']\n",
    "# datasets_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ddb3f395",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from: ensembl_transcript to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene&ids=ENST00000379939\n",
      "Edge from:ensembl_transcript:ENST00000379939, to: ncbigene:26960\n",
      "from: ensembl_transcript to: uniprot\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,uniprot&ids=ENST00000379939\n",
      "Edge from:ensembl_transcript:ENST00000379939, to: uniprot:Q5T321\n",
      "from: ensembl_transcript to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene,medgen,mondo&ids=ENST00000379939\n",
      "Edge from:ensembl_transcript:ENST00000379939, to: mondo:0030930\n",
      "from: ensembl_transcript to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene&ids=ENST00000400445\n",
      "Edge from:ensembl_transcript:ENST00000400445, to: ncbigene:26960\n",
      "from: ensembl_transcript to: uniprot\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,uniprot&ids=ENST00000400445\n",
      "Edge from:ensembl_transcript:ENST00000400445, to: uniprot:Q8NFP9-1\n",
      "from: ensembl_transcript to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene,medgen,mondo&ids=ENST00000400445\n",
      "Edge from:ensembl_transcript:ENST00000400445, to: mondo:0030930\n",
      "from: ensembl_transcript to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene&ids=ENST00000537702\n",
      "Edge from:ensembl_transcript:ENST00000537702, to: ncbigene:26960\n",
      "from: ensembl_transcript to: uniprot\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,uniprot&ids=ENST00000537702\n",
      "Edge from:ensembl_transcript:ENST00000537702, to: uniprot:Q8NFP9-3\n",
      "from: ensembl_transcript to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene,medgen,mondo&ids=ENST00000537702\n",
      "Edge from:ensembl_transcript:ENST00000537702, to: mondo:0030930\n",
      "from: ensembl_transcript to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene&ids=ENST00000629018\n",
      "from: ensembl_transcript to: uniprot\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,uniprot&ids=ENST00000629018\n",
      "Edge from:ensembl_transcript:ENST00000629018, to: uniprot:A0A0D9SF28\n",
      "from: ensembl_transcript to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ensembl_transcript,ncbigene,medgen,mondo&ids=ENST00000629018\n",
      "from: ncbigene to: ensembl_transcript\n",
      "from: ncbigene to: uniprot\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ncbigene,uniprot&ids=26960\n",
      "Edge from:ncbigene:26960, to: uniprot:Q8NFP9\n",
      "Edge from:ncbigene:26960, to: uniprot:Q5T321\n",
      "Edge from:ncbigene:26960, to: uniprot:B3KXQ8\n",
      "from: ncbigene to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=ncbigene,medgen,mondo&ids=26960\n",
      "Edge from:ncbigene:26960, to: mondo:0030930\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A0D9SF28\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A0D9SF28\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KQL6\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KQL6\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KQP5\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KQP5\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KRX1\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KRX1\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KRZ1\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KRZ1\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KS18\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KS18\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KS58\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KS58\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KSX3\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KSX3\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KTV6\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KTV6\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KTY1\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KTY1\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KWL1\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KWL1\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KX92\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KX92\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KXL4\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KXL4\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KXR7\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KXR7\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KXW8\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KXW8\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KXY8\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KXY8\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5KZ13\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5KZ13\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5QJB2\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5QJB2\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5QJQ0\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5QJQ0\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5QKR1\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5QKR1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=A0A8I5QKR6\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=A0A8I5QKR6\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=Q5T321\n",
      "Edge from:uniprot:Q5T321, to: ncbigene:26960\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=Q5T321\n",
      "Edge from:uniprot:Q5T321, to: mondo:0030930\n",
      "from: uniprot to: ensembl_transcript\n",
      "from: uniprot to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene&ids=Q8NFP9\n",
      "Edge from:uniprot:Q8NFP9, to: ncbigene:26960\n",
      "from: uniprot to: mondo\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=uniprot,ncbigene,medgen,mondo&ids=Q8NFP9\n",
      "Edge from:uniprot:Q8NFP9, to: mondo:0030930\n",
      "from: mondo to: ensembl_transcript\n",
      "from: mondo to: ncbigene\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=mondo,medgen,ncbigene&ids=0030930\n",
      "Edge from:mondo:0030930, to: ncbigene:26960\n",
      "from: mondo to: uniprot\n",
      "https://api.togoid.dbcls.jp/convert?format=json&route=mondo,medgen,ncbigene,uniprot&ids=0030930\n",
      "Edge from:mondo:0030930, to: uniprot:Q8NFP9\n",
      "Edge from:mondo:0030930, to: uniprot:Q5T321\n",
      "Edge from:mondo:0030930, to: uniprot:B3KXQ8\n"
     ]
    }
   ],
   "source": [
    "# Connecting outer nodes (!= primary dataset, the network hub node)\n",
    "for index, row in outer_entries.iterrows():\n",
    "    for target in outer_entries['dest_dataset'].unique():\n",
    "        # Loop for the datasets other than itself\n",
    "        if target != row['dest_dataset']:\n",
    "            time.sleep(1)\n",
    "            print('from: ' + row['dest_dataset'] + ' to: ' + target)\n",
    "            \n",
    "            config_dataset = datasets_dict[row['dest_dataset']]['conversion']\n",
    "            if not target in config_dataset:\n",
    "                continue\n",
    "\n",
    "            togoid_api_route = config_dataset[target]\n",
    "            api_url = togoid_api_route + row['dest_entry']\n",
    "            print(api_url)\n",
    "            \n",
    "            res_json = urlopen(api_url)\n",
    "            res = json.loads(res_json.read())\n",
    "\n",
    "            n_from = row['dest_dataset'] + ':' + row['dest_entry']\n",
    "            for i in res['results']:\n",
    "                n_to = target + ':' + i\n",
    "                print('Edge from:' + n_from + ', to: ' + n_to)\n",
    "                net.add_node(n_to)\n",
    "                net.add_edge(n_from, n_to)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "44381462",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"600px\"\n",
       "            src=\"mygraph.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x40795ce490>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "net.toggle_physics(True)\n",
    "net.show('mygraph.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f77b001d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
